crates/bevy_reflect/bevy_reflect_derive/src/impls.rs

```diff
@@ -178,6 +187,16 @@ pub(crate) fn impl_struct(derive_data: &ReflectDeriveData) -> TokenStream {
                 #partial_eq_fn
             }
         }
+
+        impl #impl_generics #bevy_reflect_path::Typed for #struct_name #ty_generics #where_clause {
+            fn type_info() -> #bevy_reflect_path::TypeInfo {
+                let fields: [#bevy_reflect_path::NamedField; #field_count] = [
+                    #(#bevy_reflect_path::NamedField::static_new::<#field_types>(#field_names),)*
```

Comment:
We should consider using `once_cell` here to make this a private lazily initialized static variable. We could return a reference here (maybe even a static one) and generally avoid recomputing / reallocating types.

Reply:
Would this work for generics? I haven't used `once_cell` all that much but I think a static in a generic impl doesn't work, right?

Reply:
Just looking at the `once_cell` documentation, I'm not sure this will be possible. We could maybe store a `HashMap<TypeId, TypeInfo>`, but there's going to be a tradeoff either way I feel.

Reply:
Ah yeah that won't work for generics. Thats a shame. I guess we could select impl code in the macro based on whether or not the type is generic :smile:. An internal static hashmap to cache each generic TypeInfo instance would probably still outperform re-generating this stuff on each access.

Reply:
Yeah that should be pretty doable. We could just create a `once_cell`'d `HashMap` for generic impls and create a `once_cell`'d `TypeInfo` otherwise.

Although would we run into performance issues when acquiring a lock on the map? I guess we only need to get a write lock when first retrieving the `TypeInfo` data 🤔